File: app/CerealExercise/CerealExercise/Services/WeeklyProgressCalculator.swift
import Foundation

enum WeeklyProgressCalculator {
    static func statuses(
        forWeekContaining date: Date,
        records: [WorkoutRecord],
        today: Date,
        rescuedDates: Set<Date> = [],
        restLimit: Int = 2,
        calendar: Calendar = .mondayFirst
    ) -> [DailyStatusEntry] {
        let week = calendar.weekInterval(containing: date)
        let restDays = Set(RestDayResolver.restDays(in: week, records: records, today: today, limit: restLimit, calendar: calendar))

        return calendar.days(in: week).map { day in
            let dayRecords = records.filter { calendar.isDate($0.date, inSameDayAs: day) }
            // 保険チケット救済日も達成扱いにするため rescuedDates を渡す。
            // これが無いと週カレンダー/週次進捗だけが救済を無視し、
            // 履歴タブの月次カレンダーや streak と食い違う (3 LLM 監査 A-Major)。
            let status = AchievementEvaluator.dailyStatus(
                for: day,
                records: records,
                restDays: restDays,
                rescuedDates: rescuedDates,
                today: today,
                calendar: calendar
            )
            return DailyStatusEntry(date: day, status: status, recordIds: dayRecords.map(\.id))
        }
    }

    static func progress(from statuses: [DailyStatusEntry]) -> WeeklyProgress {
        WeeklyProgress(
            achievedCount: statuses.filter { $0.status.countsAsAchieved }.count,
            totalDays: 7
        )
    }
}

File: app-android/app/src/main/java/com/goexercise/app/domain/WeeklyProgressCalculator.kt
package com.goexercise.app.domain

import java.time.LocalDate

/**
 * 週次進捗の計算。iOS `WeeklyProgressCalculator.swift` の 1:1 移植。
 * 週は月曜起点で常に 7 日。rescuedDates を必ず渡す(渡し忘れると週カレンダー/週次進捗
 * だけが救済を無視し、月次カレンダーや streak と食い違う再発バグ [[gotcha_rescued_dates_threading]])。
 */
object WeeklyProgressCalculator {

    fun statuses(
        weekContaining: LocalDate,
        records: List<WorkoutRecord>,
        today: LocalDate,
        rescuedDates: Set<LocalDate> = emptySet(),
        restLimit: Int = 2,
    ): List<DailyStatusEntry> {
        val weekStart = RestDayResolver.weekStart(weekContaining)
        val restDays = RestDayResolver.restDays(weekStart, records, today, restLimit).toSet()

        return RestDayResolver.weekDays(weekStart).map { day ->
            val dayRecords = records.filter { it.date == day }
            val status = AchievementEvaluator.dailyStatus(
                date = day,
                records = records,
                restDays = restDays,
                rescuedDates = rescuedDates,
                today = today,
            )
            DailyStatusEntry(date = day, status = status, recordIds = dayRecords.map { it.id })
        }
    }

    fun progress(statuses: List<DailyStatusEntry>): WeeklyProgress =
        WeeklyProgress(
            achievedCount = statuses.count { it.status.countsAsAchieved },
            totalDays = 7,
        )
}
